src/base/rss/rss_autodownloader.cpp

```diff
@@ -0,0 +1,379 @@
+        setRule_impl(rule);
+        emit ruleChanged(rule.name());
+        resetProcessingQueue();
+        m_dirty = true;
+    }
+
+    store();
+}
+
+bool AutoDownloader::renameRule(const QString &ruleName, const QString &newRuleName)
+{
+    if (!m_rules.contains(ruleName)) return false;
+    if (m_rules.contains(newRuleName)) return false;
+
+    m_rules.insert(newRuleName, m_rules.take(ruleName));
+    m_dirty = true;
+    emit ruleRenamed(newRuleName, ruleName);
+    store();
+    return true;
+}
+
+void AutoDownloader::removeRule(const QString &ruleName)
+{
+    if (m_rules.remove(ruleName) > 0) {
+        m_dirty = true;
+        emit ruleRemoved(ruleName);
+    }
+
+    store();
+}
+
+void AutoDownloader::process()
+{
+    if (m_processingQueue.isEmpty()) return; // processing was disabled
+
+    processJob(m_processingQueue.takeFirst());
+    if (!m_processingQueue.isEmpty())
+        // Schedule to process the next torrent (if any)
+        m_processingTimer->start();
+
+    store();
+}
+
+void AutoDownloader::handleTorrentDownloadFinished(const QString &url)
+{
+    auto job = m_waitingJobs.take(url);
+
+    auto feed = Session::instance()->feedByURL(job->feedURL);
+    if (feed) {
+        if (auto article = feed->articleByGUID(job->articleGUID))
+            article->markAsRead();
+    }
+}
+
+void AutoDownloader::handleTorrentDownloadFailed(const QString &url)
+{
+    m_waitingJobs.remove(url);
+    // TODO: Re-schedule job here.
+}
+
+void AutoDownloader::handleNewArticle(Article *article)
+{
+    if (!article->isRead() && !article->torrentUrl().isEmpty())
+        addJobForArticle(article);
+}
+
+void AutoDownloader::setRule_impl(const AutoDownloadRule &rule)
+{
+    m_rules.insert(rule.name(), rule);
+}
+
+void AutoDownloader::addJobForArticle(Article *article)
+{
+    const QString torrentURL = article->torrentUrl();
+    if (m_waitingJobs.contains(torrentURL)) return;
+
+    QSharedPointer<ProcessingJob> job(new ProcessingJob);
+    job->feedURL = article->feed()->url();
+    job->articleGUID = article->guid();
+    job->articleTitle = article->title();
+    job->torrentURL = torrentURL;
+    m_processingQueue.append(job);
+    if (!m_processingTimer->isActive())
+        m_processingTimer->start();
+}
+
+void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
+{
+    for (AutoDownloadRule &rule: m_rules) {
+        if (!rule.isEnabled()) continue;
+        if (!rule.feedURLs().contains(job->feedURL)) continue;
+        if (!rule.matches(job->articleTitle)) continue;
+
+//        rule.setLastMatch(QDateTime::currentDateTime());
+//        m_dirty = true;
+
+        // if rule is in ignoring state do nothing with matched torrent
+        if (rule.ignoreDays() > 0) {
+            // Optimization: update lastMatch only for rules that use it for "ignoring" purposes
+            const QDateTime lastMatch = rule.lastMatch();
+            rule.setLastMatch(QDateTime::currentDateTime());
```

Comment:
Sorry, I wrote the last comment without seeing the code (I remember it was supposed to work, so I thought that this is the case). You're right. I fixed it. Also I defer all kinds of optimization yet...
>Actually, I also just noticed that you are using currentDateTime for the match date, which is also incorrect. You should be using the date from the article.

This is legacy behavior.
Fixed now.